New chapter on important base R functions (#1113)
hadley/r4ds: base-R.qmd

```{r}
#| results: "asis"
#| echo: false
source("_common.R")
status("complete")
```

To finish off the programming section, we're going to give you a quick tour of the most important base R functions that we don't otherwise discuss in the book.
This will help you read code that you'll encounter in the wild and start your on your journey to learn different ways to tackle the same problems.

We'll focus on three big topics: subsetting with `[`, subsetting with `[[` and `$` and iteration with for loops and the apply family of functions.
These tools are particularly useful as you being to do more programming in R, which is why we don't teach them earlier.

This is a good place to remind you that the tidyverse is not only way to solve data science problems.
We teach the tidyverse in this book because it's designed as whole, which tends to increase consistency across functions making it a little easier to use.
After you read this book we expect you'll learn other ways to tackle the same problems using base R, data.table, and other packages.
You'll certainly encounter these other approaches when you start reading R code written by other people, particularly if you're using StackOverflow.

> It's 100% OK for your code to include a mix of paradigms, but when you have time, we highly recommend taking the time to rewrite with a consistent style --- this will help you understand what it does better, improve your R programming skills, and lead to code that's easier to maintain.

Reviewer: I feel like I wanted to see something here about the fact that the tidyverse would not be possible without base R.

Call it what you want, but the tidyverse is built on top of base R code (if you count the R C API as base R code, which I do).

> 

Reviewer: Maybe some reminder that you've really been using base R all along throughout this book. The entire `Transform` section uses data types and functions that come from base R, i.e. we didn't invent factors/Date objects, and we didn't invent the `&` or `>` functions.

[^base-r-1]: Read <https://adv-r.hadley.nz/subsetting.html#subset-multiple> to see how you can also subset a data frame like it's a 1d object and how you can subset it with a matrix.

Here are a couple of examples:

```{r}
df <- tibble(x = 1:3, y = c("a", "e", "f"), z = runif(3))

# Select first row and second column
df[1, 2]

# Select all rows and columns x and y
df[, c("x" , "y")]

# Select rows where `x` is greater than 1 and all columns
df[df$x > 1, ]
```

We'll come back to `$` shortly, but we need to use it here because `[` doesn't use tidy evaluation, so you need to be explicit about the source of the `x` variable.

There's an important difference between tibbles and data frames when it comes to `[`.
If `df` is a data frame, `df[, cols]` will return a vector if `col` selects a single column and a data frame if it selects more than one column.
If `df` is a tibble, then `[` will always return a tibble.

```{r}
df1 <- data.frame(x = 1:3)
df1[, "x"]

df2 <- tibble(x = 1:3)
df2[, "x"]
```

You can also avoid this problem by using 1d `[` to select columns:

```{r}
df1["x"]

> ```

Reviewer: I guess this is easier to explain than `drop = FALSE`?

Reviewer: May be worth a sentence explicitly acknowledging the difference between `x[cols]` and `x[rows,]`

    # same as
    df[order(df$x, df$y), ]
    ```

    You can use the `decreasing` argument to sort all columns in decreasing order or `-rank(col)` to individual sort column in decreasing order.

-   Both `select()` and `relocate()` are similar to subsetting the columns with a character vector.

    ```{r}
    #| results: false
    df |> select(x, z)
    # same as
    df[, c("x", "z")]
    ```

base R also provides a function that combines the features of `filter()` and `select()`[^base-r-2] called `subset()`:

[^base-r-2]: But it doesn't handle grouped data frames differently and it doesn't support the more selection helper functions like `starts_with()`.

```{r}
df |> 
  filter(x > 1) |> 
  select(y, z)

df |> subset(x > 1, c(y, z))
```

This function was the inspiration for much of dplyr's syntax.

### Exercises

1.  Create functions that take a vector as input and return:

    a.  The elements at even numbered positions.
    b.  Every element except the last value.
    c.  Only even numbers (and no missing values).

2.  Why is `x[-which(x > 0)]` not the same as `x[x <= 0]`?
    Read the documentation for `which()` and do some experiments to figure it out.

> 

Reviewer 2: Additional exercise idea:

Suppose you have the following data frame:

```r
df <- tibble(x = c(1, 2, NA, 3, 4, 5, NA))
```

Subset for rows where `x > 2` using `filter()` and also using `[`. How do the results compare?

## Selecting a single element `$` and `[[`

`[`, which selects many elements, is paired with `[[` and `$`, which extract a single element from a list or data frame.
In this section, we'll show you how to use `[[` and `$` to pull columns out of a data frames, a couple of differences between `data.frames` and tibbles, and how `[` differs from `[[` when used with lists.

### Data frames

`[[` and `$` can be used like `pull()` to extract columns out of a data frame.

> `[[` can access by position or by name, and `$` is specialized for access by position:

Reviewer: ```suggestion
`[[` can access by position or by name, and `$` is specialized for access by name:
```

```{r}
tb <- tibble(
  x = 1:4,
  y = c(10, 4, 1, 21)
)

# by position
tb[[1]]

# by name
tb[["x"]]
tb$x
```

They can also be used to create new columns, the base equivalent of `mutate()`:

```{r}
tb$z <- tb$x + tb$y
tb
```

There are a number other base approaches to creating new columns including with `transform()`, `with()`, and `within()`.
Hadley collected a few examples at <https://gist.github.com/hadley/1986a273e384fb2d4d752c18ed71bedf>.

### Tibbles

There are a couple of important differences between tibbles and base `data.frame`s when it comes to `$`.
Data frames will match the prefix of any variable names (so called **partial matching**) and don't complain if a column doesn't exist:

```{r}
df <- data.frame(x1 = 1)
df$x
df$z
```

OTOH, are more strict: they only ever match variable names exactly and they will generate a warning if the column you are trying to access doesn't exist:

```{r}
tb <- tibble(x1 = 1)

tb$x
tb$z
```

For this reason we sometimes joke that tibbles are lazy and surly: they do less and complain more.

### Lists

`[[` and `$` are also really important for working with lists, and it's really important to understand how they differ to `[`.

> Lets illustrate the differences with a list named l:

Reviewer 2: ```suggestion
Let's illustrate the differences with a list named `l`:
```

```{r}

> l <- list(a = 1:3, b = "a string", c = pi, d = list(-1, -5))

Reviewer 2: ```suggestion
l <- list(
  a = 1:3, 
  b = "a string", 
  c = pi, 
  d = list(-1, -5)
)
```

```

-   `[` extracts a sub-list.
    It doesn't matter how many elements you extract, the result will always be a list.

    ```{r}
    str(l[1:2])
    str(l[4])
    ```

    Like with vectors, you can subset with a logical, integer, or character vector.

-   `[[` and `$` extract a single component from a list.

>     They removes a level of hierarchy from the list.

Reviewer 2: ```suggestion
    They remove a level of hierarchy from the list.
```

    ```{r}
    str(l[[1]])
    str(l[[4]])

    str(l$a)
    ```

The difference between `[` and `[[` is important for lists because `[[` drills down into the list while `[` returns a new, smaller list.
This difference very important, but it's easy to get confused.
To help you remember, take a look at the an unusual pepper shaker shown in @fig-pepper-1.

> If this pepper shaker is your list `pepper`, then, `pepper[1]` is a pepper shaker containing a single pepper packet, as in @fig-pepper-2.

Reviewer 2: ```suggestion
If we suppose this pepper shaker is a list `pepper`, then, `pepper[1]` is a pepper shaker containing a single pepper packet, as in @fig-pepper-2.
```

2.  What would `pepper[[1]][1]` be?
    What about `pepper[[1]][[1]]`?

## Iteration

In @sec-iteration, you learned tidyverse techniques for iteration, primarily `dplyr::across()` and `purrr::map()`.
In this section, you'll learn about their base equivalents, first the apply family of functions and then the `for` loop, the powerful idea that underlies all these functions.

Some people will tell you to avoid `for` loops because they are slow.
They're wrong!
(Well at least they're rather out of date, as `for` loops haven't been slow for many years.) The chief benefit of using functions like `map()` or `apply()` is not speed, but clarity: once you've mastered the basic idea, they make your code faster to write and easier to read.

### Apply family

Another way of saying "map a function over each element of a vector" is "apply a function over each element of a vector", and base R provides a family of functions known as the "apply" family because they all use apply as a suffix.
Here we'll give you a quick overview of this family so you can recognize them in the wild.

The most important member of this family is `lapply()`, which is very similar to `purrr::map()`[^base-r-3].

> In fact, because we don't use any of `map()`'s more advanced features, you can replace every `map()` call in @sec-iteration with `lapply()`.

Reviewer 2: ```suggestion
In fact, because we haven't used any of `map()`'s more advanced features, you can replace every `map()` call in @sec-iteration with `lapply()`.
```

[^base-r-3]: It just lacks convenient features like progress bars and reporting which element caused the problem if there's an error.

There's no exact base R equivalent to `across()` but you can get close by using `[` twice along with `lapply():`

```{r}
df <- tibble(a = 1, b = 2, c = "a", d = "b", e = 4)

# First find numeric columns
num_cols <- sapply(df, is.numeric)
num_cols

# Then transform each column with lapply() and replace the original values
df[num_cols] <- lapply(df[num_cols], \(x) x * 2)
df
```

The code above uses a new function, `sapply()`.
It's similar to `lapply()` but it always tries to simplify the result, hence the `s` in its name, here producing a logical vector instead of a list.
We don't recommend using it for programming, because the simplification can fail and give you an unexpected type, but it's usually fine for interactive use.

> Base R provides a stricter version of `sapply()` called `vapply()`, short for vector apply.

Reviewer: Could draw some connections between `sapply->map_vec` and `vapply->map_int/dbl` if you wanted to

It takes an additional argument that specifies the expected type which ensures that simplification occurs the same way regardless of the input.

> For example, we could replace the `sapply()` call above with this `vapply()` where we specific that we expect `is.numeric()` to return a logical vector of length 1:

Reviewer 2: ```suggestion
For example, we could replace the `sapply()` call above with this `vapply()` where we specificy that we expect `is.numeric()` to return a logical vector of length 1:
```

```{r}
vapply(df, is.numeric, logical(1))
```

Another important member of the apply family is `tapply()` which computes a single grouped summary:

```{r}
diamonds |> 
  group_by(cut) |> 
  summarise(price = mean(price))

tapply(diamonds$price, diamonds$cut, mean)
```

Unfortunately it returns its results in a named vector which requires some gymnastics if you want to collect multiple summaries into a data frame matched with the variable you're grouping by.
(It's certainly possible to avoid this and just work with free floating vectors, but that generally just makes more work down the road).

> If you want to see how you might use `tapply()` or other base techniques to perform other grouped summaries, Hadley has collected a few techniques [in a gist](https://gist.github.com/hadley/c430501804349d382ce90754936ab8ec).

Reviewer: My favorite `tapply()` feature is that when given factors to "group" by, you get one summary result per factor level, even if that factor level didn't appear in the actual data. Which made it useful in hardhat for `weighted_table()`
https://github.com/tidymodels/hardhat/blob/1e6774ce9b5db35ce36a6247cd9390cd23db75da/R/table.R#L126

Somewhat like `dplyr::group_by(.drop = FALSE)`

```{r}
seq_along(paths)
```

Using the indices is important because it allows us to link to each each position in the input with the corresponding position in the output:

```{r}
for (i in seq_along(paths)) {
  files[[i]] <- readxl::read_excel(paths[[i]])
}
```

To combine the list of tibbles into a single tibble you can use `do.call()` + `rbind()`:

```{r}
do.call(rbind, files)
```

Rather than making a list and saving the results as we go, a simple approach is to build up the data frame piece-by-piece:

```{r}
out <- NULL
for (path in paths) {
  out <- rbind(out, readxl::read_excel(path))
}
```

We recommend avoiding this pattern except for short experiments because growing the data frame in this way is a fundamentally slow approach.

## Summary

<https://adv-r.hadley.nz/subsetting.html>

<https://adv-r.hadley.nz/control-flow.html#loops>

This chapter concludes the programming section of the book.
You've now learned the basics of programming in R.
You know now the data types that underpin all of the objects you work with, and have two powerful techniques (functions and iteration) for reducing the duplication in your code.

> We hope you've got a taste for how programming can help your analyses, and you've made a solid start on your journey to become not just a data scientist who uses R, but a data science who can program in R.

Reviewer 2: ```suggestion
We hope you've got a taste for how programming can help your analyses, and you've made a solid start on your journey to becoming not just a data scientist who uses R, but a data scientist who can program in R.
```

# A field guide to base R

```{r}
#| results: "asis"
#| echo: false
source("_common.R")
status("polishing")
```

To finish off the programming section, we're going to give you a quick tour of the most important base R functions that we don't otherwise discuss in the book.
These tools are particularly useful as you do more programming and will help you read code that you'll encounter in the wild.

This is a good place to remind you that the tidyverse is not only way to solve data science problems.
We teach the tidyverse in this book because tidyverse packages share a common design philosophy, which increases the consistency across functions, making each new function or package a little easier to learn and use.

> It's not possible to use the tidyverse without using base R, so we've already taught you a **lot** of base R functions: from `library()` to load packages, to `sum()` and `mean()` for numeric summaries, to the factor, date, and POSIXct data types, and of course all the basic operators like `+`, `-`, `/`, `*`, `|`, `&`, and `!`.

Reviewer 2: ```suggestion
It's not possible to use the tidyverse without using base R, so we've actually already taught you a **lot** of base R functions: from `library()` to load packages, to `sum()` and `mean()` for numeric summaries, to the factor, date, and POSIXct data types, and of course all the basic operators like `+`, `-`, `/`, `*`, `|`, `&`, and `!`.
What we haven't focused on so far is base R workflows, so we will highlight a few of those in this chapter.
```

After you read this book you'll learn other approaches to the same problems using base R, data.table, and other packages.
You'll certainly encounter these other approaches when you start reading R code written by other people, particularly if you're using StackOverflow.

> It's 100% okay to write code that uses a mixes of approaches, and don't let anyone tell you otherwise!

Reviewer 2: ```suggestion
It's 100% okay to write code that uses a mix of approaches, and don't let anyone tell you otherwise!
```
